packages/react-hooks/src/hooks/steps.ts

```diff
@@ -0,0 +1,19 @@
+import {useState, useCallback} from 'react';
+
+export function useSteps(steps: number, initialStep = 0) {
```

Comment:
Using 0 as the default initialStep might cause off by one errors for someone who is counting steps "naturally" as opposed to index-based.

```suggestion
export function useSteps(steps: number, initialStep = 1) {
```

Reply:
I was on the fence about this one too and I decided to go with 0 as the default. I think it's better to keep 0-based indexing because we can leverage them to walkthrough arrays or even objects. Maybe I could add to the documentation that it is 0-based indexed.

Reply:
I'm thinking of what would be the most common case, where people write:

```jsx
// this gives me 5 steps, not 4 as this might imply at a glance.
const {currentStep, nextStep, previousStep} = useSteps(4)

// to get exactly 4 steps I have to write
const {currentStep, nextStep, previousStep} = useSteps(4, 1)
```

Reply:
I'm convinced. Just updated it

Reply:
@mateus 
I think another win is when you are using an array of steps, it would force you to be more explicit about it:

```jsx
const {currentStep, nextStep, previousStep} = useSteps(data.length - 1, 0)
```
vs just
```jsx
const {currentStep, nextStep, previousStep} = useSteps(data.length - 1)
```

What do you think of handling arrays automatically and allowing the user to just pass them in as the first argument:

```suggestion
export function useSteps(steps: number | Array<T>, initialStep = 1) {
  let count = steps;
  if (Array.isArray(steps)) {
    count = steps.length - 1;
  }
```

So that you can write
```jsx
const {currentStep, nextStep, previousStep} = useSteps([{key: "value"}], 0)
```

Reply:
🤔   I don't see a lot of benefit in it if we only use the array to determine the number of steps. If we pass down an array of steps I would likely expect that `currentStep` isn't only a number representing the current step but also that it includes the value from the array. 

Something like this:
```tsx
const {currentStep, nextStep, previousStep} = useSteps(['potato', 'apple', 'banana'])

> currentStep 
{index: 0, value: 'potato', step: 1}

> nextStep()
> currentStep
{index: 1, value: 'apple', step: 2}
```

The scope of this hook might be growing a bit if we choose this route and I don't know if that's the right level of abstraction here. Unless I am missing something from your suggestion?

Reply:
I wanna play around with this idea a bit more though. I don't wanna discard it right away. Let me write a couple of demos using this approach and I will get back to you.

Reply:
What do you think of this:

```tsx
export function useSteps(steps: number | any[], initialStep = 1) {
  const [currentStep, setCurrentStep] = useState(initialStep);
  const stepsArr = useMemo(
    () => (Array.isArray(steps) ? steps : [...Array(steps).keys()]),
    [steps],
  );

  return {
    currentStep,
    currentValue: stepsArr[currentStep - 1],
    nextStep: useCallback(() => {
      if (currentStep < stepsArr.length) {
        setCurrentStep(currentStep + 1);
      }
    }, [currentStep, stepsArr]),
    previousStep: useCallback(() => {
      if (currentStep > 1) {
        setCurrentStep(currentStep - 1);
      }
    }, [currentStep]),
  };
}
``` 

Usage cases:

```tsx
const {currentStep, currentValue, nextStep, previousStep} = useSteps(3);

> currentStep, currentValue
1, 0

> nextStep
> currentStep, currentValue
2, 1

> nextStep
> currentStep, currentValue
3, 2
```

or

```tsx
const {currentStep, currentValue, nextStep, previousStep} = useSteps(['apple', 'banana', 'corn']);

> currentStep, currentValue
1, apple

> nextStep
> currentStep, currentValue
2, banana

> nextStep
> currentStep, currentValue
3, corn
```

or 

```tsx
const {currentStep, currentValue, nextStep, previousStep} = useSteps(['apple', 'banana', 'corn'], 2);

> currentStep, currentValue
2, banana

> nextStep
> currentStep, currentValue
3, corn
```

Reply:
I thought of the currentValue idea as well but the user of this will have access to the currentStep index so try can just do `data[currentStep]` as they need, which is why I didn’t suggest it.